src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp

```diff
@@ -334,3 +335,68 @@ ov::pass::DeReshapeMatMul::DeReshapeMatMul() {
     auto m = std::make_shared<pattern::Matcher>(final_reshape, matcher_name);
     register_matcher(m, matcher_pass_callback);
 }
+
+ov::pass::DeReshapeFullyConnected::DeReshapeFullyConnected() {
+    MATCHER_SCOPE(DeReshapeFullyConnected);
+
+    auto reshaped_input = pattern::wrap_type<op::v1::Reshape>([](Output<Node> out) -> bool {
+        const auto& input_shape = out.get_node_shared_ptr()->get_input_partial_shape(0);
+        if (input_shape.rank().is_dynamic() || input_shape.size() < 3)
+            return false;
+        const auto& output_shape = out.get_partial_shape();
+        if (output_shape.rank().is_dynamic() || output_shape.size() < 2)
+            return false;
+        return dims_are_equal(input_shape[input_shape.size() - 1], output_shape[output_shape.size() - 1]);
+    });
+    auto converted =
+        pattern::wrap_type<op::v0::Convert>({reshaped_input}, pattern::consumers_count(1));  // optional convert
+
+    auto dynamic_input = std::make_shared<pattern::op::Or>(OutputVector{reshaped_input, converted});
+    auto static_input = pattern::any_input(pattern::rank_equals(2));
+    auto mm_label = pattern::wrap_type<op::v0::MatMul>({dynamic_input, static_input}, [](Output<Node> out) -> bool {
+        auto mm = ov::as_type_ptr<op::v0::MatMul>(out.get_node_shared_ptr());
+        return mm && !mm->get_transpose_a() && pattern::consumers_count(1)(out);
+    });
+
+    auto reshaped_output =
+        pattern::wrap_type<op::v1::Reshape>({mm_label, pattern::any_input()}, [](Output<Node> out) -> bool {
+            const auto& input_shape = out.get_node_shared_ptr()->get_input_partial_shape(0);
+            if (input_shape.rank().is_dynamic() || input_shape.size() < 2)
+                return false;
+            const auto& output_shape = out.get_partial_shape();
+            if (output_shape.rank().is_dynamic() || output_shape.size() < 3)
+                return false;
+            return dims_are_equal(input_shape[input_shape.size() - 1], output_shape[output_shape.size() - 1]);
```

Comment:
It looks like very common code with `reshaped_input`, maybe is possible to parametrize it and provide expected ranks as parameters?